primitives/enclave-api/src/enclave_remote_attestation.rs

```diff
@@ -0,0 +1,260 @@
+    fn init_quote(&self) -> EnclaveResult<(sgx_target_info_t, sgx_epid_group_id_t)> {
+        let mut ti: sgx_target_info_t = sgx_target_info_t::default();
+        let mut eg: sgx_epid_group_id_t = sgx_epid_group_id_t::default();
+
+        let result = unsafe {
+            sgx_init_quote(
+                &mut ti as *mut sgx_target_info_t,
+                &mut eg as *mut sgx_epid_group_id_t,
+            )
+        };
+
+        ensure!(result == sgx_status_t::SGX_SUCCESS, Error::Sgx(result));
+
+        Ok((ti, eg))
+    }
+
+    fn calc_quote_size(&self, revocation_list: Vec<u8>) -> EnclaveResult<u32> {
+        let mut real_quote_len: u32 = 0;
+
+        let (p_sig_rl, sig_rl_size) = utils::vec_to_c_pointer_with_len(revocation_list);
+
+        let result =
+            unsafe { sgx_calc_quote_size(p_sig_rl, sig_rl_size, &mut real_quote_len as *mut u32) };
+
+        ensure!(result == sgx_status_t::SGX_SUCCESS, Error::Sgx(result));
+
+        Ok(real_quote_len)
+    }
+
+    fn get_quote(
+        &self,
+        revocation_list: Vec<u8>,
+        report: sgx_report_t,
+        quote_type: sgx_quote_sign_type_t,
+        spid: sgx_spid_t,
+        quote_nonce: sgx_quote_nonce_t,
+        quote_length: u32,
+    ) -> EnclaveResult<(sgx_report_t, Vec<u8>)> {
+        let (p_sig_rl, sig_rl_size) = utils::vec_to_c_pointer_with_len(revocation_list);
+        let p_report = &report as *const sgx_report_t;
+        let p_spid = &spid as *const sgx_spid_t;
+        let p_nonce = &quote_nonce as *const sgx_quote_nonce_t;
+
+        let mut qe_report = sgx_report_t::default();
+        let p_qe_report = &mut qe_report as *mut sgx_report_t;
+
+        let mut return_quote_buf = vec![0u8; quote_length as usize];
+        let p_quote = return_quote_buf.as_mut_ptr();
+
+        let ret = unsafe {
+            sgx_get_quote(
+                p_report,
+                quote_type,
+                p_spid,
+                p_nonce,
+                p_sig_rl,
+                sig_rl_size,
+                p_qe_report,
+                p_quote as *mut sgx_quote_t,
+                quote_length,
+            )
+        };
+
+        ensure!(ret == sgx_status_t::SGX_SUCCESS, Error::Sgx(ret));
+
+        Ok((qe_report, return_quote_buf))
+    }
+
+    fn get_update_info(
+        &self,
+        platform_blob: sgx_platform_info_t,
+        enclave_trusted: i32,
+    ) -> EnclaveResult<sgx_update_info_bit_t> {
+        let mut update_info: sgx_update_info_bit_t = sgx_update_info_bit_t::default();
+
+        let result = unsafe {
+            sgx_report_attestation_status(
+                &platform_blob as *const sgx_platform_info_t,
+                enclave_trusted,
+                &mut update_info as *mut sgx_update_info_bit_t,
+            )
+        };
+
+        ensure!(result == sgx_status_t::SGX_SUCCESS, Error::Sgx(result));
+
+        Ok(update_info)
+    }
+}
+
+impl EnclaveTlsRemoteAttestation for Enclave {
```

Comment:
I agree, however we then probably need a new name for `EnclaveApi`. Now that we have multiple API traits, it would anyway make sense to re-name it (and maybe split up into more traits?). So far we have to following methods in the `EnclaveApi` trait:
```
    fn rpc(&self, request: Vec<u8>) -> EnclaveResult<Vec<u8>>;

    /// initialize the enclave
    fn init(&self) -> EnclaveResult<()>;

    fn get_state(&self, cyphertext: Vec<u8>, shard: Vec<u8>) -> EnclaveResult<Vec<u8>>;

    fn init_chain_relay<SpHeader: Encode + Decode + Debug>(
        &self,
        genesis_header: SpHeader,
        authority_list: Vec<u8>,
        authority_proof: Vec<Vec<u8>>,
    ) -> EnclaveResult<SpHeader>;

    fn produce_blocks(&self, blocks: Vec<u8>, nonce: u32) -> EnclaveResult<()>;

    fn get_rsa_shielding_pubkey(&self) -> EnclaveResult<Rsa3072PubKey>;

    fn get_ecc_signing_pubkey(&self) -> EnclaveResult<ed25519::Public>;

    fn get_mrenclave(&self) -> EnclaveResult<[u8; MR_ENCLAVE_SIZE]>;

    fn initialize_pool(&self) -> EnclaveResult<()>;
```

What grouping would make sense to you?

* Maybe all the initialization into one trait? 
* All the key and MREnclave getters into on trait?

Reply:
Hmm, as always, this is kind of a tough question. :D

* Merging the initialization is not good because: `init_chain_relay` and `init_pool` actually comes with different business logic. I believe we even want to make `init_pool` optional.

* In a first step, we may want to have something like `EnclaveBase` trait, which contains: init, get_state, get_rsa, get, ecc, get_mrenclave.
* with business logic in mind: `initialize_pool` and `rpc` currently belongs together in a `DirectRequest` trait. Although, I am highly unsure about that. `rpc` might need to stand alone depending on the functionality it will employ in the future.
* I believe `init_chain_relay` is either standalone or included in the `EnclaveBase` as we will always need it.
* `produce_blocks` could be standalone in a `Sidechain` trait.

What do you think of that?

Reply:
* Re-named `EnclaveApi` to `EnclaveBase`
* Created `DirectRequest` trait with `rpc` and `initialize_pool` methods
* Created `SideChain` trait with `produce_blocks` method